### ISSUE TITLE
Add Inline CodeLens buttons to run tests or benchmarks

### ISSUE THREAD

**Issue by SaschaMann** [julia-vscode/julia-vscode#1307]
The Go extension has inline buttons to run and debug (individual) tests or benchmarks (see the screenshot below).

It would be nice to have that option in the Julia extension as well. It would probably make sense to have these for testsets rather than individual test cases.

I like the feature in the Go extension because I have to use the mouse to navigate to the test I run anyway, so it's a bit more convenient to just click it rather than clicking to place the cursor and then moving my hands to the keyboard to hit a shortcut/command.

![grafik](https://user-images.githubusercontent.com/20866761/83949798-0ab3ca80-a826-11ea-8db1-0eaa454fa911.png)

---

**Comment by pfitzseb** [comment]
Some comments from Slack:
```
Jesse Perla  19 hours ago
What about it it only worked on testset, which is easy to parse, and the assumption was that the user has set things up so the testset could be run (eg they had executed any includes or globals they require?)

Jesse Perla  19 hours ago
Otherwise I don't see how it would be possible... There are plenty of cases where people do the prep work in runtests.jl and then just have a bunch of testset in files.

Jesse Perla  19 hours ago
That said, I think this would be an extremely useful feature. I spend have of my time shift-enter on individual testset and then scrolling up after it runs to figure out where I was. It could even have a visual indicator at the top if the test was successful
```

**Comment by ZacLN** [comment]
This seems like a perfect use for CodeLens

**Comment by davidanthoff** [comment]
We could generally use this to run cells and codeblocks in the REPL, right? With buttons to run it normally, or in debug mode.

The test case seems tricky because Julia doesn't really seem to have a notion of reliably running a subset of tests?

**Comment by pfitzseb** [comment]
> We could generally use this to run cells and codeblocks in the REPL, right? With buttons to run it normally, or in debug mode.
>
> The test case seems tricky because Julia doesn't really seem to have a notion of reliably running a subset of tests?

Running tests (module setup code) is a subset of running code anyways. So let's just do the "Run Code", "Debug Code", "Run File", "Debug File" widgets if we do decide to implement this (I'm still not a huge fan of pushing editor content around without the user explicitly asking for it by running a command).

**Comment by davidanthoff** [comment]
We can add an option and only show these code lenses if users opt-in for now?

**Comment by ChrHorn** [comment]
Might be worth to check out the [Test Explorer UI](https://marketplace.visualstudio.com/items?itemName=hbenl.vscode-test-explorer) extension. It automatically adds code lenses for tests if you use it's api as well as more features like an explorer view and inline markers for successful/failed tests.

**Comment by davidanthoff** [comment]
Yes, agreed, we should consider adding a test adapter for Test Explorer UI. I wish that framework was part of VS Code core...
